homeassistant/components/point/__init__.py

```diff
@@ -70,62 +81,100 @@ async def async_setup(hass: HomeAssistant, config: ConfigType) -> bool:
-            DOMAIN, context={"source": config_entries.SOURCE_IMPORT}
+            DOMAIN, context={"source": SOURCE_IMPORT}, data=conf
         )
     )
 
     return True
 
 
-async def async_setup_entry(hass: HomeAssistant, entry: ConfigEntry) -> bool:
-    """Set up Point from a config entry."""
+async def async_setup_entry(hass: HomeAssistant, entry: PointConfigEntry) -> bool:
+    """Set up Minut Point from a config entry."""
+    hass.data.setdefault(DOMAIN, {})
 
-    async def token_saver(token, **kwargs):
-        _LOGGER.debug("Saving updated token %s", token)
+    if "auth_implementation" not in entry.data:
+        # Config entry is imported from old implementation without native oauth2.
         hass.config_entries.async_update_entry(
-            entry, data={**entry.data, CONF_TOKEN: token}
+            entry,
+            data={
+                **entry.data,
+                "auth_implementation": DOMAIN,
+                CONF_TOKEN: {
+                    **entry.data[CONF_TOKEN],
+                    "expires_at": 0,
+                },
+                "imported": True,
+            },
         )
 
-    session = PointSession(
-        async_get_clientsession(hass),
-        entry.data["refresh_args"][CONF_CLIENT_ID],
-        entry.data["refresh_args"][CONF_CLIENT_SECRET],
-        token=entry.data[CONF_TOKEN],
-        token_saver=token_saver,
+    implementation = (
+        await config_entry_oauth2_flow.async_get_config_entry_implementation(
+            hass, entry
+        )
+    )
+    session = config_entry_oauth2_flow.OAuth2Session(hass, entry, implementation)
+    auth = api.AsyncConfigEntryAuth(
+        aiohttp_client.async_get_clientsession(hass), session
     )
+    entry.runtime_data = auth
+
     try:
-        # the call to user() implicitly calls ensure_active_token() in authlib
-        await session.user()
-    except ConnectTimeout as err:
-        _LOGGER.debug("Connection Timeout")
+        await auth.async_get_access_token()
+    except ClientResponseError as err:
+        if err.status in {HTTPStatus.UNAUTHORIZED, HTTPStatus.FORBIDDEN}:
+            raise ConfigEntryAuthFailed from err
+        raise ConfigEntryNotReady from err
+    except ClientError as err:
         raise ConfigEntryNotReady from err
-    except Exception:  # noqa: BLE001
-        _LOGGER.error("Authentication Error")
-        return False
 
-    hass.data[DATA_CONFIG_ENTRY_LOCK] = asyncio.Lock()
-    hass.data[CONFIG_ENTRY_IS_SETUP] = set()
+    pointSession = PointSession(auth)
 
-    await async_setup_webhook(hass, entry, session)
-    client = MinutPointClient(hass, entry, session)
-    hass.data.setdefault(DOMAIN, {}).update({entry.entry_id: client})
+    client = MinutPointClient(hass, entry, pointSession)
     hass.async_create_task(client.update())
+    entry.runtime_data[DATA_CONFIG_ENTRY_CLIENT] = client
+    entry.runtime_data[DATA_CONFIG_ENTRY_LOCK] = asyncio.Lock()
+    entry.runtime_data[CONFIG_ENTRY_IS_SETUP] = set()
```

Comment:
You just assigned `auth` and now you make it behave like a dict? If you need to store more stuff in runtime_data, let's create a dataclass for that so we can do that in a typed manner

Reply:
I guess we could have a dataclass for this, yes.

Reply:
Oh, this didn't go as planned.

I created a dataclass like this:

```python
dataclass
class PointData:
    """Point Data."""

    client: MinutPointClient
    entry_lock: asyncio.Lock = asyncio.Lock()
    entries: set[str | None] = field(default_factory=set)
```
And then I changed my code to:
```suggestion
    entry.runtime_data = PointData(client)
```

But I get this error from `mypy`: `Incompatible types in assignment (expression has type "PointData", variable has type "AsyncConfigEntryAuth") `

Must the `runtime_data` store an `auth`?